ruby/sdk-lightweight.md

```diff
@@ -0,0 +1,462 @@
+    # - Positional arguments passed to activity (inline with other SDKs)
+    # - Keyword arguments are reserved for options
+    # - This call is blocking
+    workflow.execute_activity(SumActivity, a, b, start_to_close_timeout: 10)
+
+    # - Starting an activity by its name rather than a class
+    # - The result of this activity will be used as a workflow result (Ruby standard)
+    workflow.execute_activity('sum-activity', a, b, start_to_close_timeout: 10)
+  end
+end
+```
+
+#### Alternative approaches
+
+- For the same exact reasons it is not feasible to use non-namespaced methods for workflow
+
+- Similar to activities it should be possible to define multiple workflows per class, however this
+  will not be allowed to avoid a perception of a shared state between workflows and allow for
+  class-level "decorators"
+
+
+## Start a workflow
+
+```ruby
+require 'temporal/client'
+require 'sum_workflow'
+
+# Connections are eager and a lazy option will be introduced later
+connection = Temporal::Connection.connect('localhost:7233')
+client = Temporal::Client.new(connection, namespace: 'demo')
+
+# workflow is referenced by the class
+handle = client.start_workflow(SumWorkflow, 1, 2, id: 'id-1', task_queue: 'demo')
+handle.result # => 3
+```
+
+#### Alternative approaches
+
+- Starting a foreign workflow from a string name
+
+```ruby
+client.start_workflow('sum-workflow', 1, 2, id: 'id-1', task_queue: 'demo')
+```
+
+- Starting a foreign workflow with the help of a stub
+
+```ruby
+class MyWorkflow < Temporal::Workflow
+  # Optionally override the name of the workflow
+  workflow_name 'my.workflow'
+end
+
+client.start_workflow(MyWorkflow, 1, 2, id: 'id-1', task_queue: 'demo')
+```
+
+
+## Start a worker
+
+```ruby
+require 'temporal/worker'
+require 'sum_workflow'
+require 'some_other_workflow'
+require 'sum_activity'
+require 'some_other_activity'
+
+connection = Temporal::Connection.connect('localhost:7233')
+worker = Temporal::Worker.new(
+  connection,
+  namespace: 'demo',
+  task_queue: 'demo',
+  workflows: [SumWorkflow, SomeOtherWorkflow],
+  activities: [SumActivity, SomeOtherActivity]
+)
+
+worker.run
+```
+
+#### Alternative approach
+
+To support passing additional context to the activity classes a worker can be initialised with an
+instance of the activity class (only available for activities).
+
+```ruby
+redis = Redis.new('localhost:6379')
+
+Temporal::Worker.new(
+  ...,
+  activities: [CalculatorActivities.new(redis)]
+)
+```
+
+*NOTE: This however will come with a heavy warning — the activity must be strictly thread-safe as
+the provided instance will be called from multiple threads (and possibly at the same exact time).*
+
+
+## Execute a child workflow
+
+`parent_workflow.rb`
+
+```ruby
+class ChildWorkflow < Temporal::Workflow
+  def execute(name)
+    "Hello, #{name}!"
+  end
+end
+
+class ParentWorkflow < Temporal::Workflow
+  def execute
+    workflow.execute_child_workflow(ChildWorkflow, 'parent', id: 'id-2')
```

Comment:
FWIW, at Stripe, execute takes a context variable explicitly.  This makes it easier for folks to discover the stuff they're supposed to interact with.  Putting those methods on the workflow gets confusing because there are a bunch of other methods you're not supposed to use.
I'm not sure it's a good idea here without static type declarations but figured I'd mention it as an option.

Reply:
In theory, at least with all of our other SDKs, all methods on `workflow` package/module/whatever are for use in the workflow. So if they are unsure what to interact with, "just look at the workflow module". If there are methods you are not supposed to use in workflows, they shouldn't be in that module.

There is definitely a lot of discussion about the pros/cons of implicit vs explicit contexts.

Reply:
@drewhoskins-stripe not sure where the miscommunication happened, but all the public methods (without exceptions) on the `workflow` context will be available for general use. This is true in the current SDK as well.

Passing the context explicitly suggests that you would define these methods on the `Temporal::Workflow` which then has the potential of colliding with user-defined methods and also creates inconsistency in the API (calling a class vs a string name vs utility method such as `now` and `sleep`)